contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,181 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @dev Implementation of an abstract `WarpProtocolRegistry` contract.
+ *
+ * This implementation is a contract that can be used as a base contract for protocols that are
+ * build on top of warp. It allows the protocol to be upgraded through a warp out-of-band message.
+ */
+abstract contract WarpProtocolRegistry {
+    // Address that the out-of-band warp message sets as the "source" address.
+    // The address is obviously not owned by any EOA or smart contract account, so it
+    // can not possibly be the source address of any other warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    WarpMessenger public constant WARP_MESSENGER =
```

Comment:
> we allow this address to mutate with out-of-band messages

Can you clarify this? I assumed the precompile address would stay the same

Reply:
>  I assumed the precompile address would stay the same

I'm wondering if that's a safe assumption, or if we should account for that potentially changing in the future.

Reply:
discussed offline and we're going with it as a safe assumption since the precompile address is rooted in the VM, and would require a network upgrade to change the address, and there's no real benefit in changing this address, so we don't forsee it changing